File: version_3/ACO.R
library(cluster)
library(proxy)

if (!requireNamespace("aricode", quietly = TRUE)) install.packages("aricode")
library(aricode)

# -----------------------------
# Registro interno de hiperparámetros 
# Consultar: .aco_hyperparams$global / .aco_hyperparams$runs[["nombre"]]
# -----------------------------
.aco_hyperparams <- list(
  global = list(
    algorithm          = "ACO",
    distance_method    = "cosine",
    n_ants             = 50L,
    max_iterations     = 20L,
    penalty_weight     = 100,
    perturbation_rate  = 0.1,
    kmeans_nstart      = 5L,
    kmeans_iter_max    = 30L,
    adjust_cardinality_max_iter = 1000L,
    objective          = "silhouette_cosine_minus_penalty"
  ),
  runs = list()
)

compute_metrics <- function(y_true, y_predict, X) {
  y_true_int    <- as.integer(as.factor(y_true))
  y_predict_int <- as.integer(y_predict)
  ARI_val <- tryCatch(aricode::ARI(y_true_int, y_predict_int),  error = function(e) NA_real_)
  AMI_val <- tryCatch(aricode::AMI(y_true_int, y_predict_int),  error = function(e) NA_real_)
  NMI_val <- tryCatch(aricode::NMI(y_true_int, y_predict_int),  error = function(e) NA_real_)
  Sil_val <- tryCatch({
    if (length(unique(y_predict_int)) < 2) return(NA_real_)
    d   <- proxy::dist(as.matrix(X), method = "cosine")
    sil <- cluster::silhouette(y_predict_int, d)
    mean(sil[, "sil_width"], na.rm = TRUE)
  }, error = function(e) NA_real_)
  list(ARI = ARI_val, AMI = AMI_val, NMI = NMI_val, Silhouette_mean = Sil_val)
}

# Salida
dir.create("predictions", showWarnings = FALSE, recursive = FALSE)
.pred_aco_csv <- "predictions/pred_ACO.csv"

# -----------------------------
# Helpers: X numérico robusto
# -----------------------------
make_numeric_X <- function(X) {
  X <- as.data.frame(X)
  X_num <- lapply(X, function(col) {
    if (is.numeric(col) || is.integer(col)) return(as.numeric(col))
    if (is.logical(col))                    return(as.numeric(col))
    if (is.factor(col))                     return(as.numeric(col))
    if (is.character(col))                  return(as.numeric(as.factor(col)))
    return(NULL)
  })
  keep  <- !vapply(X_num, is.null, logical(1))
  X_num <- X_num[keep]
  if (length(X_num) == 0) return(NULL)
  X_num <- as.data.frame(X_num)
  X_num <- X_num[, vapply(X_num, function(z) length(unique(z)) > 1, logical(1)), drop = FALSE]
  if (ncol(X_num) == 0) return(NULL)
  X_num
}

prepare_data <- function(dataset) {
  dataset <- as.data.frame(dataset)
  if (ncol(dataset) < 2) return(NULL)
  if ("class" %in% colnames(dataset)) {
    y <- dataset$class
    X <- dataset[, setdiff(colnames(dataset), "class"), drop = FALSE]
  } else {
    y <- dataset[[ncol(dataset)]]
    X <- dataset[, -ncol(dataset), drop = FALSE]
  }
  X_num <- make_numeric_X(X)
  if (is.null(X_num)) return(NULL)
  list(X = X_num, y = as.factor(y))
}

# -----------------------------
# CORRECCIÓN 1: calculate_centroids
# -----------------------------
calculate_centroids <- function(X, cluster_assignment, k) {
  X <- as.data.frame(X)
  centroids_df <- aggregate(X, by = list(cluster = cluster_assignment), FUN = mean)
  centroids_df <- centroids_df[order(centroids_df$cluster), , drop = FALSE]
  centroids    <- as.matrix(centroids_df[, -1, drop = FALSE])
  
  if (nrow(centroids) < k) {
    tmp        <- matrix(NA_real_, nrow = k, ncol = ncol(centroids))
    tmp[as.integer(centroids_df$cluster), ] <- centroids
    centroids  <- tmp
  }
  if (any(is.na(centroids))) {
    global_mu <- colMeans(as.matrix(X), na.rm = TRUE)
    for (r in seq_len(nrow(centroids))) {
      if (any(is.na(centroids[r, ]))) centroids[r, ] <- global_mu
    }
  }
  centroids
}

# -----------------------------
# CORRECCIÓN 2: adjust_cardinality
# OPT: tabulate en vez de sapply para contar tamaños
# -----------------------------
adjust_cardinality <- function(cluster_assignment, X, centroids, target_cardinality) {
  max_iterations <- 1000
  iteration      <- 0
  k              <- length(target_cardinality)
  # OPT: tabulate es O(n) en C, reemplaza sapply(1:k, function(c) sum(...))
  cluster_sizes  <- tabulate(cluster_assignment, nbins = k)
  
  while (any(cluster_sizes > target_cardinality) && iteration < max_iterations) {
    iteration <- iteration + 1
    
    for (j in which(cluster_sizes > target_cardinality)) {
      idx <- which(cluster_assignment == j)
      if (length(idx) == 0L) next
      
      element   <- idx[1]
      distances <- colSums((t(centroids) - as.numeric(X[element, ]))^2)
      
      available_clusters <- which(cluster_sizes < target_cardinality)
      valid_clusters     <- setdiff(available_clusters, j)
      
      if (length(valid_clusters) == 0L) {
        cluster_assignment[element] <- -1L
        cluster_sizes[j]            <- cluster_sizes[j] - 1L
      } else {
        chosen                       <- valid_clusters[which.min(distances[valid_clusters])]
        cluster_assignment[element]  <- chosen
        cluster_sizes[j]             <- cluster_sizes[j] - 1L
        cluster_sizes[chosen]        <- cluster_sizes[chosen] + 1L
      }
    }
  }
  
  unassigned_idx <- which(cluster_assignment == -1L)
  if (length(unassigned_idx) > 0L) {
    for (element in unassigned_idx) {
      available_clusters <- which(cluster_sizes < target_cardinality)
      if (length(available_clusters) > 0L) {
        chosen                      <- available_clusters[which.min(cluster_sizes[available_clusters])]
        cluster_assignment[element] <- chosen
        cluster_sizes[chosen]       <- cluster_sizes[chosen] + 1L
      } else {
        cluster_assignment[element] <- 1L
        cluster_sizes[1]            <- cluster_sizes[1] + 1L
      }
    }
  }
  
  if (iteration >= max_iterations) {
    warning("adjust_cardinality alcanzó el límite de iteraciones.")
  }
  
  cluster_assignment
}

# -----------------------------
# ACO principal
# OPT: dist coseno NxN se calcula una sola vez y se pasa a evaluate_solution
# OPT: penalty vectorizada con tabulate
# OPT: perturbación vectorizada sin loop punto por punto
# -----------------------------
run_ACO <- function(X, target_cardinality) {
  set.seed(123)
  Xmat <- as.matrix(X)
  n    <- nrow(Xmat)
  k    <- length(target_cardinality)
  
  if (is.na(k) || k < 2)  stop("k inválido (<2).")
  if (k >= n)              stop(paste0("k=", k, " inválido para n=", n, "."))
  
  n_ants         <- 50
  max_iterations <- 20
  penalty_weight <- 100
  
  # OPT: calcular dist coseno NxN una sola vez (antes: se recalculaba en cada evaluate_solution)
  d_cosine <- proxy::dist(Xmat, method = "cosine")
  
  # Función objetivo: silhouette coseno - penalización por violación
  # OPT: recibe d_cosine pre-calculada
  evaluate_solution <- function(cluster_assignment) {
    if (length(unique(cluster_assignment)) < 2) return(-Inf)
    ss <- tryCatch(cluster::silhouette(cluster_assignment, d_cosine), error = function(e) NULL)
    if (is.null(ss)) return(-Inf)
    # OPT: tabulate + sum vectorizado reemplaza for-loop de penalty
    current_counts <- tabulate(cluster_assignment, nbins = k)
    penalty <- penalty_weight * sum(abs(current_counts - target_cardinality))
    mean(ss[, "sil_width"]) - penalty
  }
  
  # Solución inicial: kmeans + ajuste heurístico simple
  generate_initial_solution <- function() {
    km                 <- kmeans(Xmat, centers = k, nstart = 5, iter.max = 30)
    cluster_assignment <- km$cluster
    for (j in seq_len(k)) {
      while (sum(cluster_assignment == j) > target_cardinality[j]) {
        idx                            <- which(cluster_assignment == j)
        cluster_assignment[sample(idx, 1)] <- sample(setdiff(1:k, j), 1)
      }
    }
    cluster_assignment
  }
  
  # Perturbación: cambia ~10% de los puntos aleatoriamente
  # Loop original preservado: runif(1) por punto mantiene la secuencia exacta del RNG
  perturb_solution <- function(cluster_assignment) {
    new_ca <- cluster_assignment
    for (j in 1:n) {
      if (runif(1) < 0.1) new_ca[j] <- sample(1:k, 1)
    }
    new_ca
  }
  
  best_score             <- -Inf
  best_cluster_assignment <- NULL
  
  for (iteration in seq_len(max_iterations)) {
    ants <- vector("list", n_ants)
    for (i in seq_len(n_ants)) {
      ca       <- generate_initial_solution()
      ca       <- perturb_solution(ca)
      ants[[i]] <- ca
    }
    for (i in seq_len(n_ants)) {
      score <- evaluate_solution(ants[[i]])
      if (is.finite(score) && score > best_score) {
        best_score              <- score
        best_cluster_assignment <- ants[[i]]
      }
    }
  }
  
  if (!is.null(best_cluster_assignment)) {
    centroids               <- calculate_centroids(Xmat, best_cluster_assignment, k)
    best_cluster_assignment <- adjust_cardinality(
      best_cluster_assignment,
      Xmat,
      centroids,
      target_cardinality
    )
  }
  
  list(best_solution = best_cluster_assignment, best_score = best_score)
}

# -----------------------------
# Runner: genera una fila de resultado por dataset
# -----------------------------
run_clustering_row <- function(dataset, target_cardinality, dataset_name) {
  data <- prepare_data(dataset)
  if (is.null(data)) return(NULL)
  X <- data$X
  y <- data$y
  if (is.null(target_cardinality) || length(target_cardinality) < 2 || all(is.na(target_cardinality))) return(NULL)
  
  set.seed(123)
  start_total <- proc.time()
  
  results    <- tryCatch(run_ACO(X, as.integer(target_cardinality)), error = function(e) NULL)
  
  if (is.null(results) || is.null(results$best_solution)) return(NULL)
  
  y_predict <- results$best_solution
  
  total_time <- (proc.time() - start_total)[3]
  
  metrics <- compute_metrics(y, y_predict, X)
  
  # Guardar hiperparámetros de esta ejecución (solo en memoria)
  .aco_hyperparams$runs[[dataset_name]] <<- list(
    k                 = length(target_cardinality),
    n                 = length(y_predict),
    n_features        = ncol(as.matrix(X)),
    target_sizes      = as.integer(target_cardinality),
    n_ants            = 50L,
    max_iterations    = 20L,
    penalty_weight    = 100,
    perturbation_rate = 0.1,
    distance_method   = "cosine"
  )
  
  data.frame(
    name = dataset_name,
    n = length(y_predict),
    k = length(unique(y_predict)),
    y_predict = paste(as.integer(y_predict), collapse = " "),
    y_true = paste(as.integer(y), collapse = " "), 
    target_cardinality = paste(as.integer(target_cardinality), collapse = " "),
    cardinality_pred   = paste(as.integer(table(factor(y_predict,
                                                       levels = 1:length(target_cardinality)))), collapse = " "),
    Execution_Time = as.numeric(total_time),
    ARI = metrics$ARI,
    AMI = metrics$AMI,
    NMI = metrics$NMI,
    Silhouette_mean = metrics$Silhouette_mean,
    stringsAsFactors = FALSE
  )
}

# -----------------------------
# Loop principal
# OPT: pre-asignar lista, seq_len, vapply
# -----------------------------
results_list <- vector("list", nrow(odatasets_unique))

for (i in seq_len(nrow(odatasets_unique))) {
  cat("\n\n--- Executing ACO for dataset at position:", i, "---\n")
  
  tryCatch({
    dataset            <- odatasets_unique$dataset[[i]]
    dataset_name       <- odatasets_unique$name[[i]]
    target_cardinality <- odatasets_unique$class_distribution_vector[[i]]
    
    if (is.null(target_cardinality) || all(is.na(target_cardinality)) || length(target_cardinality) < 2) {
      cat("Target cardinality inválida. Skipping.\n")
      next
    }
    
    row_result <- run_clustering_row(dataset, target_cardinality, dataset_name)
    
    if (!is.null(row_result)) {
      results_list[[i]] <- row_result
      cat("Time:", row_result$Execution_Time, "s\n")
    } else {
      cat("Dataset skipped.\n")
    }
    
  }, error = function(e) {
    cat("Error:", e$message, "\n")
  })
}

results_clean <- results_list[!vapply(results_list, is.null, logical(1))]

if (length(results_clean) > 0) {
  write.table(do.call(rbind, results_clean), .pred_aco_csv,
              sep = ",", row.names = FALSE, col.names = TRUE)
  cat("\nACO guardado en:", .pred_aco_csv, "\n")
} else {
  cat("\nACO finalizado sin resultados válidos. No se generó CSV.\n")
}

File: python/algorithms/aco.py
"""
algorithms/aco.py
=================
ACO (Ant Colony Optimization) para clustering con cardinalidad fija.
Migración Python del archivo ACO.R.

Estructura del algoritmo:
  - 50 hormigas, 20 iteraciones
  - Cada hormiga genera una solución NUEVA en cada iteración (kmeans + perturbación)
  - Función objetivo: silhouette (coseno) - penalty * |violación de cardinalidad|
    (penalty_weight = 100, mucho más agresivo que BAT que usa 10)
  - Al final aplica adjust_cardinality a la mejor solución encontrada

Diferencia con R:
  - El comentario en ACO.R dice: "Loop original preservado: runif(1) por punto
    mantiene la secuencia exacta del RNG" (no vectorizaron la perturbación
    para mantener fidelidad RNG entre versiones de R).
  - En Python ya asumimos que el RNG es distinto a R, así que VECTORIZAMOS
    la perturbación. Esto da una mejora notable de velocidad: en lugar de N
    iteraciones del bucle, una sola operación con máscara booleana.
"""

from __future__ import annotations

import logging
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Optional

import numpy as np
import pandas as pd
from scipy.spatial.distance import pdist, squareform
from sklearn.cluster import KMeans

from ._common import (
    adjust_cardinality,
    calculate_centroids,
    compute_metrics,
    evaluate_solution,
    prepare_data,
    tabulate_clusters,
    to_int_list,
)

logger = logging.getLogger(__name__)
if not logger.handlers:
    logging.basicConfig(level=logging.INFO, format="%(message)s")


# ============================================================================
# Hiperparámetros y rutas de salida
# ============================================================================

@dataclass
class AcoHyperparams:
    """Hiperparámetros globales del algoritmo ACO."""
    algorithm:                   str   = "ACO"
    distance_method:             str   = "cosine"
    n_ants:                      int   = 50
    max_iterations:              int   = 20
    penalty_weight:              float = 100.0
    perturbation_rate:           float = 0.1
    kmeans_nstart:               int   = 5
    kmeans_iter_max:             int   = 30
    adjust_cardinality_max_iter: int   = 1000
    master_seed:                 int   = 123
    objective:                   str   = "silhouette_cosine_minus_penalty"


ACO_HYPERPARAMS: dict[str, Any] = {
    "global": AcoHyperparams(),
    "runs":   {},
}

PROJECT_ROOT    = Path(__file__).resolve().parent.parent
PREDICTIONS_DIR = PROJECT_ROOT / "predictions"
PREDICTIONS_DIR.mkdir(parents=True, exist_ok=True)
PRED_ACO_CSV    = PREDICTIONS_DIR / "pred_ACO.csv"


# ============================================================================
# Solución inicial: K-means + ajuste random simple
# ============================================================================

def _generate_initial_solution(
    x_mat: np.ndarray,
    target_cardinality: np.ndarray,
    rng: np.random.Generator,
    kmeans_nstart: int = 5,
    kmeans_iter_max: int = 30,
) -> np.ndarray:
    """
    Solución inicial: K-means + ajuste heurístico simple.

    A diferencia de BAT (que usa centroides para mover el punto al cluster
    más cercano), aquí ACO usa ajuste random: si un cluster está sobre-poblado,
    toma un punto al azar y lo manda a un cluster random distinto.
    """
    k = int(len(target_cardinality))

    # KMeans con random_state derivado del RNG para reproducibilidad encadenada
    km = KMeans(
        n_clusters=k,
        n_init=kmeans_nstart,
        max_iter=kmeans_iter_max,
        random_state=int(rng.integers(0, 2**31 - 1)),
    )
    cluster_assignment: np.ndarray = km.fit_predict(x_mat).astype(int) + 1

    target = np.asarray(target_cardinality, dtype=int)
    other_clusters_cache = {
        j: np.array([c for c in range(1, k + 1) if c != j], dtype=int)
        for j in range(1, k + 1)
    }

    for j in range(1, k + 1):
        while int(np.sum(cluster_assignment == j)) > int(target[j - 1]):
            idx = np.where(cluster_assignment == j)[0]
            if len(idx) == 0:
                break
            chosen_point = int(rng.choice(idx))
            new_cluster  = int(rng.choice(other_clusters_cache[j]))
            cluster_assignment[chosen_point] = new_cluster

    return cluster_assignment


# ============================================================================
# Perturbación vectorizada (mejora vs R)
# ============================================================================

def _perturb_solution(
    cluster_assignment: np.ndarray,
    k: int,
    perturbation_rate: float,
    rng: np.random.Generator,
) -> np.ndarray:
    """
    Cambia ~`perturbation_rate` de los puntos a un cluster aleatorio.

    R usa un loop punto por punto con runif(1) para preservar la secuencia
    del RNG. En Python vectorizamos: una máscara booleana + asignación masiva.
    """
    new_ca = cluster_assignment.copy()
    n      = int(len(new_ca))
    mask   = rng.random(n) < perturbation_rate
    if mask.any():
        new_ca[mask] = rng.integers(1, k + 1, size=int(mask.sum()))
    return new_ca


# ============================================================================
# ACO (loop principal)
# ============================================================================

def run_aco_algorithm(
    x_mat: np.ndarray,
    target_cardinality: np.ndarray,
    n_ants:            int   = 50,
    max_iterations:    int   = 20,
    penalty_weight:    float = 100.0,
    perturbation_rate: float = 0.1,
    kmeans_nstart:     int   = 5,
    kmeans_iter_max:   int   = 30,
    master_seed:       int   = 123,
) -> dict[str, Any]:
    """ACO con caché de distancia coseno y perturbación vectorizada."""
    rng   = np.random.default_rng(master_seed)
    x_mat = np.asarray(x_mat, dtype=float)
    n     = int(x_mat.shape[0])
    k     = int(len(target_cardinality))

    if k < 2:
        raise ValueError("k inválido (<2).")
    if k >= n:
        raise ValueError(f"k={k} inválido para n={n}.")

    # OPT: distancia coseno N×N una sola vez (ya estaba en R)
    d_cosine_sq = squareform(pdist(x_mat, metric="cosine"))

    best_score:    float                = float("-inf")
    best_solution: Optional[np.ndarray] = None

    for _ in range(max_iterations):
        # En cada iteración se genera una población NUEVA (sin memoria
        # entre iteraciones, a diferencia de BAT). Por eso ACO es más caro.
        ants: list[np.ndarray] = []
        for _ in range(n_ants):
            ca = _generate_initial_solution(
                x_mat, target_cardinality, rng,
                kmeans_nstart=kmeans_nstart,
                kmeans_iter_max=kmeans_iter_max,
            )
            ca = _perturb_solution(ca, k, perturbation_rate, rng)
            ants.append(ca)

        for ant in ants:
            score = evaluate_solution(
                ant, d_cosine_sq, target_cardinality, penalty_weight
            )
            if np.isfinite(score) and score > best_score:
                best_score    = score
                best_solution = ant.copy()

    # Ajuste final de cardinalidad sobre la mejor solución (igual que R)
    if best_solution is not None:
        centroids     = calculate_centroids(x_mat, best_solution, k)
        best_solution = adjust_cardinality(
            best_solution, x_mat, centroids, target_cardinality
        )

    return {
        "best_solution": best_solution,
        "best_score":    best_score,
    }


# ============================================================================
# Runner por dataset
# ============================================================================

def run_clustering_row(
    dataset: pd.DataFrame,
    target_cardinality: list[int],
    dataset_name: str,
) -> Optional[dict[str, Any]]:
    """Ejecuta ACO sobre un dataset y devuelve la fila de resultados."""
    data = prepare_data(dataset)
    if data is None:
        return None

    x_df:     pd.DataFrame    = data["X"]
    y_factor: pd.Categorical = data["y"]

    target = np.asarray(target_cardinality, dtype=int)
    x_mat  = x_df.to_numpy(dtype=float)

    start_total = time.perf_counter()

    try:
        result = run_aco_algorithm(x_mat, target)
    except Exception as e:  # pylint: disable=broad-exception-caught
        logger.warning("Error en run_aco_algorithm para %s: %s", dataset_name, e)
        return None

    if result.get("best_solution") is None:
        return None

    y_pred     = np.asarray(result["best_solution"], dtype=int)
    y_int      = (y_factor.codes + 1).astype(int)
    total_time = time.perf_counter() - start_total

    metrics = compute_metrics(y_int, y_pred, x_mat)

    ACO_HYPERPARAMS["runs"][dataset_name] = {
        "k":                  len(target),
        "n":                  len(y_pred),
        "n_features":         x_mat.shape[1],
        "target_sizes":       target.tolist(),
        "n_ants":             50,
        "max_iterations":     20,
        "penalty_weight":     100.0,
        "perturbation_rate":  0.1,
        "distance_method":    "cosine",
    }

    cardinality_pred = tabulate_clusters(y_pred, len(target))

    return {
        "name":               dataset_name,
        "n":                  len(y_pred),
        "k":                  int(len(np.unique(y_pred))),
        "y_predict":          " ".join(str(int(c)) for c in y_pred),
        "y_true":             " ".join(str(int(c)) for c in y_int),
        "target_cardinality": " ".join(str(int(c)) for c in target),
        "cardinality_pred":   " ".join(str(int(c)) for c in cardinality_pred),
        "Execution_Time":     float(total_time),
        "ARI":                metrics["ARI"],
        "AMI":                metrics["AMI"],
        "NMI":                metrics["NMI"],
        "Silhouette_mean":    metrics["Silhouette_mean"],
    }


# ============================================================================
# Punto de entrada para testing.py
# ============================================================================

def run(odatasets_unique: pd.DataFrame) -> None:
    """Loop principal equivalente al de ACO.R."""
    results: list[dict[str, Any]] = []

    for i in range(len(odatasets_unique)):
        print(
            f"\n--- Executing ACO for dataset at position: {i + 1} ---",
            flush=True,
        )
        try:
            dataset      = odatasets_unique["dataset"].iloc[i]
            dataset_name = odatasets_unique["name"].iloc[i]
            target_raw   = odatasets_unique["class_distribution_vector"].iloc[i]

            target = to_int_list(target_raw)
            if target is None or len(target) < 2:
                print("Target cardinality inválida. Skipping.")
                continue

            row = run_clustering_row(dataset, target, dataset_name)

            if row is not None:
                results.append(row)
                print(f"Time: {row['Execution_Time']:.4f} s")
            else:
                print("Dataset skipped.")

        except Exception as e:  # pylint: disable=broad-exception-caught
            print(f"Error: {e}")

    if results:
        df_out = pd.DataFrame(results)
        df_out.to_csv(PRED_ACO_CSV, index=False)
        print(f"\nACO guardado en: {PRED_ACO_CSV}")
    else:
        print("\nACO finalizado sin resultados válidos. No se generó CSV.")
